File: csharp/sdk/Sdk/Utils/Amount.cs
using System.Numerics;

namespace Acc.Service.Sdk.Utils;

/// <summary>
///     The class that defines the Amount
/// </summary>
public class Amount(BigInteger value, uint decimals = 18)
{
    public BigInteger Value { get; } = value;
    public uint Decimals { get; } = decimals;

    public static Amount Make(int pAmount, uint decimals = 18)
    {
        return Make(pAmount.ToString(), decimals);
    }

    public static Amount Make(string value, uint decimals = 18)
    {
        if (value.Equals("")) return new Amount(new BigInteger(0), decimals);
        value = value.Replace(",", "").Replace("_", "");
        var zeroString = "";
        for (var idx = 0; idx < decimals; idx++)
            zeroString += "0";
        var numbers = value.Split('.');
        if (numbers.Length == 1) return new Amount(BigInteger.Parse(numbers[0] + zeroString), decimals);
        var pointString = numbers[1];
        if (pointString.Length > decimals) pointString = pointString[..(int)decimals];
        else if (pointString.Length < decimals) pointString = pointString.PadRight((int)decimals, '0');
        var integral = BigInteger.Parse(numbers[0] + zeroString);
        return new Amount(BigInteger.Add(integral, BigInteger.Parse(pointString)), decimals);
    }

    public string ToAmountString()
    {
        var factor = BigInteger.Pow(new BigInteger(10), (int)Decimals);
        var integral = BigInteger.Divide(Value, factor);
        var decimals = BigInteger.Subtract(Value, BigInteger.Multiply(integral, factor));
        var integralString = integral.ToString();
        var decimalsString = decimals.ToString();
        if (decimalsString.Length < Decimals) decimalsString = decimalsString.PadLeft((int)Decimals, '0');
        return integralString + "." + decimalsString;
    }

    public override string ToString()
    {
        return Value.ToString();
    }
}

File: java/sdk/src/main/java/org/kios/service/sdk/utils/Amount.java
package org.kios.service.sdk.utils;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.math.BigInteger;

public class Amount {
    private final BigInteger _value;
    private final int _decimals;


    public Amount(BigInteger value) {
        this(value, 18);
    }

    public Amount(BigInteger value, int decimals) {
        this._value = value;
        this._decimals = decimals;
    }

    public BigInteger getValue() {
        return this._value;
    }

    public int getDecimals() {
        return this._decimals;
    }

    @NotNull
    @Contract("_, _ -> new")
    public static Amount make(String value, int decimals) {
        if (value.isEmpty()) return new Amount(BigInteger.valueOf(0), decimals);
        value = value.replace(",", "").replace("_", "");
        String Zero = "0".repeat(Math.max(0, decimals));
        String [] numbers = value.split("\\.");
        if (numbers.length == 1) return new Amount(new BigInteger(numbers[0] + Zero, 10));
        String strDecimals = numbers[1];
        if (strDecimals.length() > decimals) strDecimals = strDecimals.substring(0, decimals);
        else if (strDecimals.length() < decimals) {
            int n = decimals - strDecimals.length();
            strDecimals = strDecimals + "0".repeat(Math.max(0, decimals - strDecimals.length()));
        }
        BigInteger n1 = new BigInteger(numbers[0] + Zero, 10);
        BigInteger n2 = new BigInteger(strDecimals, 10);

        return new Amount(n1.add(n2), decimals);
    }

    @NotNull
    @Contract("_ -> new")
    public static Amount make(String value) {
        return Amount.make(value, 18);
    }

    public String toString() {
        return this._value.toString();
    }

    public String toAmountString() {
        BigInteger factor = BigInteger.valueOf(10).pow(this._decimals);
        BigInteger integral = this._value.divide(factor);
        BigInteger decimal = this._value.subtract(integral.multiply(factor));
        String integral_string = integral.toString();
        String decimals_string = decimal.toString();
        if (decimals_string.length() < this._decimals) {
            decimals_string = "0".repeat(Math.max(0, this._decimals - decimals_string.length())) + decimals_string;
        }
        return integral_string + "." + decimals_string;
    }
}
